
# Introduction

In this case study, I will be analyzing a public dataset using SQL for Bellabeat, a high-tech manufacturer of health-focused products for women. I'll be looking at smart device data to gain insight on how consumers use their smart devices and provide recommendations to the Bellabeat marketing strategy.

# Background

Bellabeat was founded in 2013 by Urška Sršen and Sando Mur with the goal of developing beautifully designed technology that would inform and inspire women. The technology would collect data on activity, sleep, stress, and reproductive health to empower women with knowledge about their own health.

Bellabeat products are available at a number of online retailers in addition to their website. The company has invested in traditional advertising media such as radio, billboards, print, and television, but focuses on digital marketing extensively. They have ads on Youtube and Google, and are active on multiple social media platforms including Facebook, Instagram, and Twitter.

**Bellabeat Products**
- Bellabeat app provides users with health data related to their activity, sleep, stress, menstrual cycle, and mindfulness habits. This data can help users better understand their current habits and make healthy decisions. The Bellabeat app connects to their line of smart wellness products.
- Leaf: A wellness tracker that can be worn as a bracelet, necklace, or clip and connects to the Bellabeat app to track activity, sleep, and stress.
- Time: A wellness watch with smart technology and connects to the Bellabeat app to track user activity, sleep, and stress.
- Spring: A water bottle with smart technology and connects to the Bellabeat app to track daily water intake.
- Bellabeat membership: A subscription-based membership program for users to have 24/7 access to fully personalized guidance on nutrition, activity, sleep, health and beauty, and mindfulness based on their lifestyle and goals.



### Business Task

Analyze smart device data to gain insight on how consumers use smart devices and answer the following questions:

- What are some trends in smart device usage?
- How could these trends apply to Bellabeat customers?
- How could these trends help influence Bellabeat marketing strategy?



### Prepare Data

The dataset was obtained from [Kaggle](https://www.kaggle.com/datasets/arashnic/fitbit) and contains FitBit tracker data from 30 users including minute, hourly, and daily level output data for activity intensity, steps, calories, sleep, and heart rate. The dataset contained 18 CSV files, each file containing a table varying in number of columns and content. Using Pandas and a Python loop, each file was read to a dataframe then written to a PostgreSQL database as a new table. 


In [2]:
#Environment setup
import pandas as pd
import psycopg2
import sqlalchemy
%load_ext sql

#Connect to SQL database
connect_url = 'postgresql://{}:{}@{}:{}/{}'.format(user,password,host,port,db)
engine = sqlalchemy.create_engine(connect_url)
%sql $connect_url
%config SqlMagic.displaycon = False

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [4]:
#Read multiple CSV files and load into SQL database
import glob 
import os
file_names = glob.glob('data/*.csv')

for names in file_names:
    tablename = os.path.basename(names)
    tablename, ext = os.path.splitext(tablename)
    df = pd.read_csv(names)
    df.columns = df.columns.str.lower() #convert column names to lower case
    df.to_sql(tablename, engine, if_exists='replace', index=False)

PendingRollbackError: Can't reconnect until invalid transaction is rolled back.  Please rollback() fully before proceeding (Background on this error at: https://sqlalche.me/e/20/8s2b)

In [3]:
%%sql
-- Verify tables in database
SELECT table_name
FROM information_schema.tables
WHERE table_schema = 'public';

0 rows affected.


[]

### Process Data
I verified that a table was sucessfully created in the database with each CSV file. I found that there was a large number of tables to explore and wanted to simplify the dataset by eliminating tables with redundant information or combine those with similar content.

<p align="center">
    <img src='images\databasediagram.png' width=70%>
</p>

- The `dailycalories`, `dailyintensities`, and `dailysteps` tables where dropped as the data already existed in the `dailyactivity` table. 
- A new table with hourly data was created by joining the `hourlycalories`, `hourlyintensities`, and `hourlysteps` tables; the 3 tables were subsequently dropped.
- Tables containing minute data were dropped as the data is too granular to extract any meaningful insights. 

In summary, the following tables were dropped:
* `dailycalories`
* `dailyintensities`
* `dailysteps`
* `hourlycalories`
* `hourlyintensities`
* `hourlysteps`
* `minutecaloriesnarrow`
* `minutecalorieswide`
* `minuteintensitiesnarrow`
* `minuteintensitieswide`
* `minutemetsnarrow`
* `minutesleep`
* `minutestepsnarrow`
* `minutestepswide`
* `heartrate_seconds`
* `weightloginfo`



In [5]:
%%sql
-- Join hourly tables and create new table
CREATE TABLE hourlydata AS(
SELECT c.id,
       c.activityhour,
       c.calories,
       i.totalintensity,
       i.averageintensity,
       s.steptotal
FROM hourlycalories AS c
FULL OUTER JOIN hourlyintensities AS i
    ON c.id = i.id
    AND c.activityhour = i.activityhour
FULL OUTER JOIN hourlysteps AS s
    ON i.id = s.id
    AND i.activityhour = s.activityhour
);

 * postgresql://postgres:***@localhost:5432/bellabeat
22099 rows affected.


[]

In [6]:
%%sql
-- Drop tables not used
DROP TABLE IF EXISTS dailycalories, dailyintensities, dailysteps, hourlycalories, hourlyintensities, hourlysteps, minutecaloriesnarrow, minutecalorieswide, minuteintensitiesnarrow, minuteintensitieswide, minutemetsnarrow, minutesleep, minutestepsnarrow, minutestepswide, heartrate_seconds, weightloginfo;

 * postgresql://postgres:***@localhost:5432/bellabeat
Done.


[]

In [7]:
%%sql
-- Verify tables were dropped
SELECT table_name
FROM information_schema.tables
WHERE table_schema = 'public';

 * postgresql://postgres:***@localhost:5432/bellabeat
3 rows affected.


[('dailyactivity',), ('sleepday',), ('hourlydata',)]

I proceeded to clean the data
- In `sleepday` table, the `sleepday` column was renamed to sleepdate to avoid confusion.
- In `sleepday` table, the `sleepdate` column was changed to a date data type.
- In `dailyactivity` table, the `activitydate` column was changed to a date data type.
- In `hourlydata` table, the `activityhour` column was changed to a timestamp data type.
- Tables were checked for duplicate rows
    - In `sleepday` table, 3 duplicate rows were found and deleted



In [8]:
%%sql
--Rename sleepday column
ALTER TABLE sleepday
RENAME COLUMN sleepday TO sleepdate;
--Change activitydate type to date
ALTER TABLE dailyactivity
ALTER COLUMN activitydate TYPE date USING activitydate::date;
--Change sleepdate type to date
ALTER TABLE sleepday
ALTER COLUMN sleepdate TYPE date USING sleepdate::date;
--Change activityhour type to timestamp
ALTER TABLE hourlydata
ALTER COLUMN activityhour TYPE timestamp USING activityhour::timestamp;

 * postgresql://postgres:***@localhost:5432/bellabeat
Done.
Done.
Done.
Done.


[]

In [ ]:
%%sql
-- Review table columns and data types
SELECT c.table_name,
       c.column_name,
       c.data_type
FROM information_schema.columns c
JOIN information_schema.tables t
ON c.table_name = t.table_name
WHERE c.table_schema = 'public'
ORDER BY table_name, column_name;

 * postgresql://postgres:***@localhost:5432/bellabeat
26 rows affected.


[('dailyactivity', 'activitydate', 'date'),
 ('dailyactivity', 'calories', 'bigint'),
 ('dailyactivity', 'fairlyactiveminutes', 'bigint'),
 ('dailyactivity', 'id', 'bigint'),
 ('dailyactivity', 'lightactivedistance', 'double precision'),
 ('dailyactivity', 'lightlyactiveminutes', 'bigint'),
 ('dailyactivity', 'loggedactivitiesdistance', 'double precision'),
 ('dailyactivity', 'moderatelyactivedistance', 'double precision'),
 ('dailyactivity', 'sedentaryactivedistance', 'double precision'),
 ('dailyactivity', 'sedentaryminutes', 'bigint'),
 ('dailyactivity', 'totaldistance', 'double precision'),
 ('dailyactivity', 'totalsteps', 'bigint'),
 ('dailyactivity', 'trackerdistance', 'double precision'),
 ('dailyactivity', 'veryactivedistance', 'double precision'),
 ('dailyactivity', 'veryactiveminutes', 'bigint'),
 ('hourlydata', 'activityhour', 'timestamp without time zone'),
 ('hourlydata', 'averageintensity', 'double precision'),
 ('hourlydata', 'calories', 'bigint'),
 ('hourlydata', 'id', 'bigint'),
 ('hourlydata', 'steptotal', 'bigint'),
 ('hourlydata', 'totalintensity', 'bigint'),
 ('sleepday', 'id', 'bigint'),
 ('sleepday', 'sleepdate', 'date'),
 ('sleepday', 'totalminutesasleep', 'bigint'),
 ('sleepday', 'totalsleeprecords', 'bigint'),
 ('sleepday', 'totaltimeinbed', 'bigint')]

In [9]:
%%sql
--Check for duplicates
SELECT id,
       sleepdate,
       COUNT(*)
FROM sleepday
GROUP BY id, sleepdate
ORDER BY COUNT(*) DESC;

SELECT id,
       activitydate,
       COUNT(*)
FROM dailyactivity
GROUP BY id, activitydate
ORDER BY COUNT(*) DESC;

SELECT id,
       activityhour,
       COUNT(*)
FROM hourlydata
GROUP BY id, activityhour
ORDER BY COUNT(*) DESC;

--Duplicate rows deleted
DELETE
FROM sleepday
WHERE ctid IN (SELECT ctid
               FROM (SELECT ctid,
                            ROW_NUMBER() OVER (PARTITION BY id, sleepday) AS rn
                     FROM sleepday) AS temptable
               WHERE rn > 1);

 * postgresql://postgres:***@localhost:5432/bellabeat
410 rows affected.
940 rows affected.
22099 rows affected.
3 rows affected.


[]

In [10]:
%%sql
--View dailyactivity table
SELECT *
FROM dailyactivity
LIMIT 5;

 * postgresql://postgres:***@localhost:5432/bellabeat
5 rows affected.


[(1503960366, datetime.date(2016, 4, 12), 13162, 8.5, 8.5, 0.0, 1.87999999523163, 0.550000011920929, 6.05999994277954, 0.0, 25, 13, 328, 728, 1985),
 (1503960366, datetime.date(2016, 4, 13), 10735, 6.96999979019165, 6.96999979019165, 0.0, 1.57000005245209, 0.689999997615814, 4.71000003814697, 0.0, 21, 19, 217, 776, 1797),
 (1503960366, datetime.date(2016, 4, 14), 10460, 6.73999977111816, 6.73999977111816, 0.0, 2.44000005722046, 0.400000005960464, 3.91000008583069, 0.0, 30, 11, 181, 1218, 1776),
 (1503960366, datetime.date(2016, 4, 15), 9762, 6.28000020980835, 6.28000020980835, 0.0, 2.14000010490417, 1.25999999046326, 2.82999992370605, 0.0, 29, 34, 209, 726, 1745),
 (1503960366, datetime.date(2016, 4, 16), 12669, 8.15999984741211, 8.15999984741211, 0.0, 2.71000003814697, 0.409999996423721, 5.03999996185303, 0.0, 36, 10, 221, 773, 1863)]

In [11]:
%%sql
--View sleepday table
SELECT *
FROM sleepday
LIMIT 5;

 * postgresql://postgres:***@localhost:5432/bellabeat
5 rows affected.


[(1503960366, datetime.date(2016, 4, 12), 1, 327, 346),
 (1503960366, datetime.date(2016, 4, 13), 2, 384, 407),
 (1503960366, datetime.date(2016, 4, 15), 1, 412, 442),
 (1503960366, datetime.date(2016, 4, 16), 2, 340, 367),
 (1503960366, datetime.date(2016, 4, 17), 1, 700, 712)]

In [12]:
%%sql
--View hourlydata table
SELECT *
FROM hourlydata
LIMIT 5;

 * postgresql://postgres:***@localhost:5432/bellabeat
5 rows affected.


[(1503960366, datetime.datetime(2016, 4, 12, 0, 0), 81, 20, 0.333333, 373),
 (1503960366, datetime.datetime(2016, 4, 12, 1, 0), 61, 8, 0.133333, 160),
 (1503960366, datetime.datetime(2016, 4, 12, 2, 0), 59, 7, 0.116667, 151),
 (1503960366, datetime.datetime(2016, 4, 12, 3, 0), 47, 0, 0.0, 0),
 (1503960366, datetime.datetime(2016, 4, 12, 4, 0), 48, 0, 0.0, 0)]

### Analyze Data

After cleaning the data, I connected the dataset to Tableau to build visualizations and extract insights. My Tableau workbook can be found here: <br>
[https://public.tableau.com/views/bellabeat_dashboard_16815197263150/Dashboard1?:language=en-US&:display_count=n&:origin=viz_share_link](https://public.tableau.com/views/bellabeat_dashboard_16815197263150/Dashboard1?:language=en-US&:display_count=n&:origin=viz_share_link)
<br>
<br>

<img src='images\userdays_bar.png' width=60%>

This bar chart illustrates the number of days each user tracked their activity. The majority of users tracked their activity for at least 30 days. Only a subset of 22 users tracked their sleep activity and only 13 users tracked their sleep for more than 15 days.

<br>
<br>

<img src='images\activity_pie.png' width=60%>
<img src='images\activity_scatter.png' width=60%>

On average, most users remained sedentary for 81% of the day. The scatterplot demonstrates a positive correlation between the number of steps taken and the number of calories burned. 

<br>
<br>

<img src='images\steps_calorie_line.png' width=60%>
<img src='images\heatmap.png' width=60%>

We found that users are most active between 5PM and 7PM, and least active between 12AM and 5AM. However, on Saturday, there is an increase in user activity between 11AM and 2PM.


### Share Results
- What are some trends in smart device usage?
    - Most users tracked their activity for at least 30 days, but only 13 users tracked their sleep for more than 15 days.
    - Users are most active between 5PM and 7PM, and least active between 12AM and 5AM. However, on Saturday, there is an increase in user activity between 11AM and 2PM.
    - On average, most users remained sedentary for 81% of the day. 
    - The scatterplot demonstrates a positive correlation between the number of steps taken and the number of calories burned.
- How could these trends apply to Bellabeat customers?
    - It seems that a majority of users are not using their smart devices to track their sleep. The device might not be comfortable to wear to bed, the device may need to be taken off to charge the battery, or the user may not be aware of the sleep tracking freature. Bellabeat could create a marketing campaign to educate users on the sleep tracking feature and encourage them to track their sleep.
    - We find that most users remain sedentary for most of the day. Bellabeat could create a marketing campaign to encourage users to be more active throughout the day with reminders to meet daily step goals. 
